Nos/Service/RelayService.swift

```diff
@@ -85,13 +86,53 @@ final class RelayService: WebSocketDelegate, ObservableObject {
     
 	func requestEvents(from client: WebSocketClient, filter: Filter = Filter()) {
         do {
-            let request: [Any] = ["REQ", UUID().uuidString, filter.dictionary]
+            // Track this so we can close requests if needed
+            filter.uuid = UUID().uuidString
+            let request: [Any] = ["REQ", filter.uuid, filter.dictionary]
             let requestData = try JSONSerialization.data(withJSONObject: request)
             let requestString = String(data: requestData, encoding: .utf8)!
             print(requestString)
             client.write(string: requestString)
         } catch {
-            print("could not send request \(error.localizedDescription)")
+            print("Error: Could not send request \(error.localizedDescription)")
+        }
+    }
+    
+    func requestEventsFromAll(filter: Filter = Filter()) {
+        // Keep this open
+        openSocketsForRelays()
+
+        // Ignore redundant requests
+        guard !requestQueue.contains(filter) else {
+            print("📡Request already open. Ignoring. \(requestQueue.count) outstanding requests")
+            return
+        }
+        
+        // TODO: This only works because all requests are of kinds [.text, .metaData, .contactList] together
```

Comment:
Ah, I see now why you are batching up all the requests in one Filter. I hooked this branch up to a big relay and it wasn't loading anyone's name or profile picture, I think because the subscription was really just sending us events. So I think we'll need a different strategy for deduplication.

Here's one that I'm interested to hear your feedback on. It doesn't really do deduplication at all in RelayService, but I think requests will naturally deduplicate if we tie them more closely to what the user is looking at:

There are two types of Filters, single event filters and subscriptions.
- When the app launches, we use a bunch of single event Filters to refresh the metadata and contact lists of ourself and everyone we are following.
- When we display a NoteCard with no author metadata, we use a single event filter to fetch their metadata
- Everything else is a long-running subscriptions, but we should tie the lifetime of a subscription to the lifetime of the view that displays that data. For example:
    - When the HomeFeedView appears (`.onAppear`) we open a subscription for all .text events for everyone we follow. When it disappears (`.onDisappear`) we close that filter.
    - When we open the ThreadView, we open a subscription for all events that reference the root note.
    - When we open a profile page we can do a one-off refresh of metadata and contact list, then open a subscription to all text events for that user.
 
With this scheme I think we can always stay under the 10 subscription limit, because the user will only be looking at a few screens at a time and our remaining slots can probably iterate through single event Filters really quickly. What do you think? CC @jcswift23 @shanecb also.

Reply:
So I guess the simple rule would boil down to this: only open a subscription for the screen the user is currently looking at. Everything else should be a one-time filter.